src/v/cluster/partition_balancer_planner.cc

```diff
@@ -0,0 +1,385 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cluster/partition_balancer_planner.h"
+
+#include "cluster/health_monitor_types.h"
+#include "cluster/scheduling/constraints.h"
+#include "cluster/scheduling/types.h"
+#include "cluster/types.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+
+#include <utility>
+
+namespace cluster {
+
+partition_balancer_planner::partition_balancer_planner(
+  ss::sharded<topic_table>& topic_table,
+  ss::sharded<partition_allocator>& partition_allocator,
+  planner_config&& config)
+  : _config(std::move(config))
+  , _topic_table(topic_table)
+  , _partition_allocator(partition_allocator) {}
+
+void partition_balancer_planner::clear_status() {
+    _ntp_sizes.clear();
+    _node_reports.clear();
+    _assigned_reallocation_sizes.clear();
+    _unavailable_nodes.clear();
+    _nodes_with_disk_constraints_violation.clear();
+    _cur_batch_size = 0;
+}
+
+void partition_balancer_planner::
+  init_ntp_sizes_and_node_reports_from_health_report(
+    const cluster_health_report& health_report) {
+    for (const auto& node_report : health_report.node_reports) {
+        _node_reports[node_report.id] = node_report;
+        for (const auto& tp_ns : node_report.topics) {
+            for (const auto& partition : tp_ns.partitions) {
+                _ntp_sizes[model::ntp(
+                  tp_ns.tp_ns.ns, tp_ns.tp_ns.tp, partition.id)]
+                  = partition.size_bytes;
+            }
+        }
+    }
+    return;
+}
+
+static bool
+node_disk_space_sorter(node_disk_space const& lhs, node_disk_space const& rhs) {
+    return lhs.free_space_percent < rhs.free_space_percent;
+}
+
+void partition_balancer_planner::
+  calculate_nodes_with_disk_constraints_violation() {
+    for (const auto& n : _node_reports) {
+        for (const auto& disk : n.second.local_state.disks) {
+            double available_space = double(disk.free) / double(disk.total);
+            if (available_space < double(1) - _config.max_disk_usage()) {
+                _nodes_with_disk_constraints_violation.emplace_back(
+                  node_disk_space{
+                    .node_id = n.second.id,
+                    .free_space_percent = available_space,
+                    .free_space = disk.free,
+                    .total_space = disk.total});
+            }
+        }
+    }
+    std::sort(
+      _nodes_with_disk_constraints_violation.begin(),
+      _nodes_with_disk_constraints_violation.end(),
+      &node_disk_space_sorter);
+}
+
+void partition_balancer_planner::calculate_unavailable_nodes(
```

Comment:
Can we use `cluster_health_report` here to check which node is unavailable ? We soon are going to move the health manager not to use raft heartbeats but pings on its own. You may use `members_table` to get node list and health reports to determine if node is up or not

Reply:
The problem with health reports is that they don't include ping timestamp (only the `is_alive` flag) and the timeout for moving the partitions is much bigger than the liveness timeout for follower metrics, so it seemed easier to use the raft0 heartbeat timestamp directly. If the health report will use its own pings and include the timestamps, we can use that.